### ISSUE TITLE
Add `pam_umask` for logins

### ISSUE THREAD

**Issue by 0n-s** [NixOS/nixpkgs#346922]
## Issue description
The default umask from `/etc/login.defs` is not applied for user logins because there's nothing that does so. The usual mechanism for this is `pam_umask`. This is not configured on NixOS as of writing; the specific proposal here is to add it to the relevant files in `/etc/pam.d`.

NB: this will inadvertently be a breaking change that should be mentioned in release notes. Reason being that the __intended default__ NixOS umask (again, see `/etc/login.defs`) is more restrictive than the __effective default__ 0022 that everyone gets right now. The more restrictive umask is better for single-user systems, so it should probably be kept as the default.

### Steps to reproduce
1. Set a non-`0022` umask in `/etc/login.defs` (the current default `077` works fine)
2. login to a user account
3. `umask` in a shell gives 0022

## Technical details

Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.
```
 - system: `"x86_64-linux"`
 - host os: `Linux 6.11.0, NixOS, 24.11 (Vicuna), 24.11pre688563.bc947f541ae5`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.18.8`
 - channels(root): `"nixos"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`
```

**Comment by 0n-s** [comment]
@majiir (as maintainer of `nixos/modules/security/pam.nix`)

**Comment by Majiir** [comment]
Makes sense. Switching to `077` could be disruptive to users that are inadvertently relying on `022` in scripts/etc, so I think we should consider putting this change behind a `stateVersion` check, even though it is a bugfix. It would be difficult for users to check their configs for any possible breakage just by reading a release note. We can add an option so that users on an old `stateVersion` can upgrade safely.

**Comment by crasm** [comment]
Considering it's been the de-facto default, I think the default value of `security.loginDefs.settings.UMASK` should be changed to 022. Then as long as `pam_umask.so` is added to `/etc/pam.d/login`, any explicit setting would be honored (if I'm not mistaken). I can see a umask of 077 being set in the hardened profile, for example.

**Comment by 0n-s** [comment]
022 is not so much about hardening as it is about public and private systems.

022 is from a time where individual computers were much more of a communal resource, and if you made a file in a publicly accessible place, you probably wanted others to be able to read it too, by default. Given the way in which the vast majority of people use computers has changed so dramatically, I’d say that behavior is only likely to cause overly lax permissions nowadays.

**Comment by Majiir** [comment]
From `login.defs(5)`:

> ### `UMASK` (number)
>
> The file mode creation mask is initialized to this value. If not specified, the mask will be initialized to 022.
>
> `useradd` and `newusers` use this mask to set the mode of the home directory they create if `HOME_MODE` is not set.
>
> It is also used by `pam_umask` as the default umask value.

In NixOS, `HOME_MODE` is _not_ set. It isn't even a documented option under `security.loginDefs.settings`. But there's a hint in the [`UMASK` option](https://github.com/NixOS/nixpkgs/blob/e8b9d879ce13b37915bf3b7988976fdbbfe45c9c/nixos/modules/programs/shadow.nix#L141) definition: `# Ensure privacy for newly created home directories` when setting the default to `077`.

So, we can clean up the mess and avoid breaking anything by doing this:

1. Add a `security.loginDefs.settings.HOME_MODE` option.
2. Default `HOME_MODE` to `077`, matching the current default for new home directories.
3. Default `UMASK` to `022`, matching the current effective default for the file mode creation mask.
4. Enable `pam_umask` by default, having no effect unless the user changes their `UMASK` setting from the default.

Then, we can have a separate discussion about changing the `UMASK` default to `077`. But since that won't be blocking any bug fixes, I doubt it will be a priority.

[EDIT] I guess if someone manually enabled `pam_umask` and wants `077`, then step (3) would be a breaking change for them. Or if someone configured `UMASK` expecting it to apply to new home directories, but did not configure `HOME_MODE`, then step (2) would be a breaking change. I think these are small enough edge cases, but we should make a breaking changes announcement.
